### DISCUSSION TITLE
Trouble with Android and the config plugin feature

### DISCUSSION THREAD

**Discussion by aciereszko** [invertase/react-native-firebase#5794] [Q&A] [answered]
Hi,

I've been looking to integrate this:

https://github.com/invertase/react-native-firebase/pull/5480

From @barthap into our app. It looks awesome, but I'm running into problems getting Crashlytics to run on Android. On iOS its fine.

Unless I'm missing something, it seems that the setup for Android differs from iOS only in the google-services.json file and its inclusion in the app.json file.

```
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./app/assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "package": "...",
      "versionCode": 5,
      "googleServicesFile": "./android/google-services.json"
    }
```

The author mentions that the setup is the same as the with-rn-firebase plugin, which also mentions the addition of the installAndroid option, but that didn't do anything.

So while on iOS it all works, on Android I get this error:

```
[Fri Oct 15 2021 10:02:48.870]  LOG      Running "main" with {"rootTag":1}
[Fri Oct 15 2021 10:02:48.871]  ERROR    Error: No Firebase App '[DEFAULT]' has been created - call firebase.initializeApp()
[Fri Oct 15 2021 10:02:48.872]  ERROR    Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.
```

@barthap or anyone else maybe knows what I could be missing?

Thanks!

Andrzej

**Comment by barthap** [comment, marked as answer]
The configuration in `app.json` for the "official" plugin difers a bit from the one from the old `with-rn-firebase` plugin.

Have you included both `@react-native-firebase/app` and `@react-native-firebase/crashlytics` plugins into the `plugins` section of `app.json` [as described here](https://rnfirebase.io/#installation-1)?

**Reply by aciereszko** [reply]
Hi Bartek,

Thanks for the quick reaction. I did yes... 

```
    "plugins": [
      "@react-native-firebase/app",
      "@react-native-firebase/crashlytics"
    ]  
```

and it works fine on iOS, just Android throws the error. I see that the appropriate firebase code got generated in the AppDelegate.m is there anything analogous to check for on the Android side? Also the error message indicates that firebase.initializeApp needs to be called, but that's a setup for the web based firebase SDK no?

**Reply by barthap** [reply]
Yeah, there are 3 things to check:
- If `google-services.json` is copied into the `android/app` directory
- If `android/build.gradle` contains google/firebase dependencies inside `buildscript` section
- If `android/app/build.gradle` contains google/firebase `apply plugin` lines at the very bottom

**Reply by aciereszko** [reply]
> Yeah, there are 3 things to check:
> 
> * If `google-services.json` is copied into the `android/app` directory
> * If `android/build.gradle` contains google/firebase dependencies inside `buildscript` section
> * If `android/app/build.gradle` contains google/firebase `apply plugin` lines at the very bottom

Oh great, so 1 and 3 weren't there... I figured out it was my misconfiguration in the end. I had the google-services.json in a different place, and even though the path was reflected in app.json correctly, I guess some part of the process wasn't triggered and app/buiild.gradle wasn't updated. That would explain the error message I was getting with the dependencies failing to load maybe?

Anyway, once I followed the directions exactly, cleared the yarn cash, installed everything back, did expo prebuild --clean and reran the prebuild, it worked.

Thanks @barthap, that helped a lot.

**Reply by aciereszko** [reply]
Also @barthap, correct me if I'm wrong, but the analytics has not been pluginized yet right? I can't follow the same pattern with `@react-native-firebase/analytics` right?

**Reply by mikehardy** [reply]
I do not believe there is anything to do at the native layer for analytics? There are some firebase.json items but those should work even without an expo plugin

**Reply by aciereszko** [reply]
> I do not believe there is anything to do at the native layer for analytics? There are some firebase.json items but those should work even without an expo plugin

Oh great, same with firestore I guess? Just follow the general directions: https://rnfirebase.io/firestore/usage ?

But then this would be using the firebase web SDK right? So I would have two configs? One with plist and json file for ios and android for crashlytics, and then the web SDK config for analytics and firestore?

Thanks guys for the knowledge dump.

**Reply by mikehardy** [reply]
I'm not aware of anything with firestore either, no - it should "just work". As long as you are using `@react-native-firebase/nnnn` packages, you're using the firebase-ios-sdk and firebase-android-sdk under the covers

**Reply by aciereszko** [reply]
I see, but it wouldn't work in the expo app -> managed workflow right?

When including this import: 

```
import firestore from "@react-native-firebase/firestore";
```

I'm getting this error when running expo start:

```
Invariant Violation: Native module cannot be null.
at node_modules/react-native/Libraries/Core/ExceptionsManager.js:104:6 in reportException
at node_modules/react-native/Libraries/Core/ExceptionsManager.js:171:19 in handleException
at node_modules/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at node_modules/expo-error-recovery/build/ErrorRecovery.fx.js:12:21 in ErrorUtils.setGlobalHandler$argument_0

Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.
at node_modules/react-native/Libraries/Core/ExceptionsManager.js:104:6 in reportException
at node_modules/react-native/Libraries/Core/ExceptionsManager.js:171:19 in handleException
at node_modules/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at node_modules/expo-error-recovery/build/ErrorRecovery.fx.js:12:21 in ErrorUtils.setGlobalHandler$argument_0
```

which goes away if I remove the import. The import is fine when running expo run:ios or when running directly from Xcode

**Reply by barthap** [reply]
> I see, but it wouldn't work in the expo app -> managed workflow right?

Right, Expo Go app doesn't support custom native code, it is mentioned [in the docs](https://rnfirebase.io/#expo). So you need to build a [custom development client](https://docs.expo.dev/clients/introduction/) to be able to use RNFB in a managed-workflow way. [Learn more here](https://docs.expo.dev/workflow/customizing/).

Btw. I wrote a tutorial on integrating RNFB + Expo Dev Client [here](https://bartlomiej-klocek.medium.com/how-to-integrate-react-native-firebase-into-expo-d34712eaf64d).

**Reply by estebanfdiazp** [reply]
Hey @barthap do you have a GitHub repository of the RNFB + Expo Dev Client tutorial ? I want to use RNFB database and Firestore but it does not seem to work. I don't really know what I am doing wrong.

**Reply by barthap** [reply]
~~No, but I'm considering updating my very old RNFB demo repo to use Dev Client~~

**Edit:** I updated my old CRUD repo and got it working with SDK 44, RNFB and Dev Client. You can find it here: https://github.com/barthap/ProductManagerDemo/pull/18 - the code is available on the [`sdk44-dev-client` branch](https://github.com/barthap/ProductManagerDemo/tree/sdk44-dev-client). Not everything is updated yet (UI bugs, analytics had breaking changes meanwhile), but it should be mostly working as expected. The demo utilizes RNFB app, auth, firestore and analytics modules. Details can be found in the PR description or in readme (still partially outdated)

**Reply by barthap** [reply]
Ok, I updated [the demo project](https://github.com/barthap/ProductManagerDemo) to use Expo SDK 44, Dev Client and React Native Firebase (app, firestore, auth and analytics). It still has some minor issues, but thy're unrelated. The app works on both platforms as expected.

**Reply by mikehardy** [reply]
@barthap thank you!
